traitsui/qt4/datetime_editor.py

```diff
@@ -15,60 +15,79 @@
 """
 
 
-import datetime
-
-from pyface.qt import QtGui
+from pyface.qt import QtGui, qt_api
 from pyface.qt.QtCore import QDateTime
-from traits.api import Instance
+from traits.api import Datetime, on_trait_change
 
 from .editor import Editor
 from .editor_factory import ReadonlyEditor as BaseReadonlyEditor
 
 
+
 class SimpleEditor(Editor):
     """ Simple Traits UI time editor that wraps QDateTimeEdit.
     """
 
     #: the earliest datetime allowed by the editor
-    minimum_datetime = Instance(datetime.datetime)
+    minimum_datetime = Datetime
 
     #: the latest datetime allowed by the editor
-    maximum_datetime = Instance(datetime.datetime)
+    maximum_datetime = Datetime
 
     def init(self, parent):
         """ Finishes initializing the editor by creating the underlying toolkit
             widget.
         """
-        self.control = QtGui.QDateTimeEdit()
+        # set min and max early, don't wait for editor sync
         self.minimum_datetime = self.factory.minimum_datetime
         self.maximum_datetime = self.factory.maximum_datetime
-        self.update_date_range()
+
+        self.control = QtGui.QDateTimeEdit()
         self.control.dateTimeChanged.connect(self.update_object)
+        self.update_date_range()
 
     def update_editor(self):
         """ Updates the editor when the object trait changes externally to the
             editor.
         """
         value = self.value
-        if value:
-            q_datetime = QDateTime(value)
-            self.control.setTime(q_datetime)
+        if self.value:
+            if self.minimum_datetime and self.minimum_datetime > value:
+                value = self.minimum_datetime
+            elif self.maximum_datetime and self.value > self.maximum_datetime:
+                value = self.maximum_datetime
+            try:
+                q_datetime = QDateTime(value)
+            except Exception:
+                pass
+            self.control.setDateTime(q_datetime)
 
     def update_object(self, q_datetime):
         """ Handles the user entering input data in the edit control.
         """
-        self.value = q_datetime.toPyDateTime()
-
+        try:
+            if qt_api == 'pyside2':
+                self.value = q_datetime.toPython()
+            else:
+                self.value = q_datetime.toPyDateTime()
+        except ValueError:
+            pass
+
+    @on_trait_change('minimum_datetime,maximum_datetime')
     def update_date_range(self):
         if self.control is not None:
             if self.minimum_datetime is not None:
                 self.control.setMinimumDateTime(
                    QDateTime(self.minimum_datetime)
                 )
+            else:
+                self.control.clearMinimumDateTime()
```

Comment:
Are users expected to mutate the `minimum_datetime` and `maximum_datetime` after the editor is created, i.e. after `init`?  If they are expected to be mutated, it is going to get quite messy e.g. the user sets the minimum value to something larger than the current value.  If the users are not expected to mutate the range, then the `clearMinimumDateTime` call is a bit redundant.  Or did I miss something?

Is it now too late to make these two variables private?
(Again, this may be a separate issue)

Reply:
It's not heavily exposed, but the values can be accessed via the editor, and if we are going to have them, we should properly support them being mutated.  Because a valid value is `None` we use that as a signal for "use the Qt defaults" which is what `clearMinimumDate` does.

The Qt editor handles all of the gorpiness of incompatible ranges (it will automatically adjust the value and upper bound to be consistent if one or other is lower than the actual value).

However there _does_ need to either be a feedback loop so that changes to the min and max in Qt get reflected in the Traits, _or_ we need a sanity check for the min and max on the Python side before we mess with things.  We already have the feedback loop for the value.

Reply:
Agreed that there is a need to synchronize the min and max in case Qt adapted them, but that part is out-of-scope here.

Confirmed the values are sychronized, with these tests, which are passing now:
<details>
    <summary> Tests when min and max are mutated </summary>

```

    def test_datetime_editor_mutate_minimum_datetime_after_init(self):
        # Test if the minimum_datetime on the editor is mutated after init,
        # the values on the edited object and the view are synchronized.
        editor_factory = DatetimeEditor(
            maximum_datetime=datetime.datetime(2000, 1, 1),
        )
        view = get_date_time_simple_view(editor_factory)
        instance = InstanceWithDatetime()
        with self.launch_editor(instance, view) as editor:

            # This value is in-range
            instance.date_time = datetime.datetime(1999, 1, 1)

            # Now change the editor's maximum datetime such that the original
            # value is out-of-range
            new_maximum_datetime = datetime.datetime(1900, 1, 1)
            editor.maximum_datetime = new_maximum_datetime

            # does not seem needed to flush the event loop, but just in case.
            self.gui.process_events()

            displayed_value = to_datetime(editor.control.dateTime())

        self.assertEqual(instance.date_time, displayed_value)
        self.assertEqual(displayed_value, new_maximum_datetime)

    def test_datetime_editor_mutate_minimum_datetime_after_init(self):
        # Test if the minimum_datetime on the editor is mutated after init,
        # the values on the edited object and the view are synchronized.
        editor_factory = DatetimeEditor(
            minimum_datetime=datetime.datetime(2000, 1, 1),
        )
        view = get_date_time_simple_view(editor_factory)
        instance = InstanceWithDatetime()
        with self.launch_editor(instance, view) as editor:

            # This value is in-range
            instance.date_time = datetime.datetime(2001, 1, 1)

            # Now change the editor's minimum datetime such that the original
            # value is out-of-range
            new_minimum_datetime = datetime.datetime(2010, 1, 1)
            editor.minimum_datetime = new_minimum_datetime

            # does not seem needed to flush the event loop, but just in case.
            self.gui.process_events()

            displayed_value = to_datetime(editor.control.dateTime())

        self.assertEqual(instance.date_time, displayed_value)
        self.assertEqual(displayed_value, new_minimum_datetime)
```

</details>